rascaline/src/calculators/soap/spherical_expansion.rs

```diff
@@ -198,38 +188,60 @@ impl SphericalExpansion {
 
 
                 if let Some(ref contribution_gradients) = contribution.gradients {
-                    if let Some(ref mut positions_gradients) = result.positions_gradients_by_pair {
+                    if let Some(ref mut positions_gradients) = result.positions_gradient_by_pair {
                         let gradients = &mut positions_gradients.slice_mut(s![pair_id, .., .., ..]);
                         gradients.assign(contribution_gradients);
                     }
 
-                    if let Some(ref mut positions_gradients) = result.positions_gradients_self {
-                        let mut gradients = positions_gradients.slice_mut(s![neighbor_type_i, mapped_center, .., .., ..]);
-                        gradients -= contribution_gradients;
+                    if pair.first != pair.second {
+                        if let Some(ref mut positions_gradients) = result.self_positions_gradients {
+                            let mut gradients = positions_gradients.slice_mut(s![neighbor_type_i, mapped_center, .., .., ..]);
+                            gradients -= contribution_gradients;
+                        }
                     }
 
                     if let Some(ref mut cell_gradients) = result.cell_gradients {
                         let mut cell_gradients = cell_gradients.slice_mut(
                             s![neighbor_type_i, mapped_center, .., .., .., ..]
                         );
 
-                        for spatial_1 in 0..3 {
-                            for spatial_2 in 0..3 {
-                                let inverse_cell_pair_vector_2 = inverse_cell_pair_vector[spatial_2];
+                        for abc in 0..3 {
+                            let shift = pair.cell_shift_indices[abc] as f64;
+                            for xyz in 0..3 {
```

Comment:
I think before you are using `xyz_1` and `xyz_2` and here seems like `abc` and `xyz`. Maybe we should be consistent.

Reply:
this is intentional: the components are called `xyz_1` and `xyz_2` for strain gradients, but `abc` and `xyz` for the cell gradients.

Reply:
Not yes, that's something I need to add. They match https://lab-cosmo.github.io/metatensor/latest/atomistic/outputs.html, but should also be documented here.

Reply:
Okay are we doing this here or in another PR?

Reply:
I'll open an issue and do this in a separate PR =)